File: dotnetBackEnd/dotnetBackend/Services/PlayerService.cs
using Backend.Data;
using Domain;
using Microsoft.EntityFrameworkCore;
using Shared;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Backend.Services
{
    public class PlayerService : IPlayerService
    {
        private readonly DbSet<Player> players;
        private readonly DbSet<Country> countries;
        private readonly Context context;

        public PlayerService(Context context)
        {
            this.context = context;
            players = context.Players;
            countries = context.Countries;
        }

        public async Task<Player> CreatePlayerAsync(PlayerDto.Create player)
        {
            if (player == null)
                return null;

            var exists = await players
                .Include(p => p.Country)
                .SingleOrDefaultAsync(p =>
                    p.Name == player.Name &&
                    p.Country.Code == player.CountryCode
                ) == null ? false : true;

            if(!exists)
            {
                var country = await countries.SingleOrDefaultAsync(c => c.Code == player.CountryCode);
                var newPlayer = new Player(
                    player.PlayerID, 
                    player.Name, 
                    player.Gender,
                    country,
                    player.Active,
                    player.Birthyear,
                    player.PlayHand,
                    player.PlayStyle,
                    player.Grip
                );
                if (country == null)
                {
                    var newCountry = new Country(player.CountryCode);
                    newPlayer.Country = newCountry;
                    await countries.AddAsync(newCountry);
                }
                else
                    newPlayer.Country = country;
                await players.AddAsync(newPlayer);
            }

            await context.SaveChangesAsync();

            return await players
                .Include(p => p.Country)
                .SingleOrDefaultAsync(p => p.Name == player.Name && p.Country.Code == player.CountryCode);
        }

        public async Task<bool> DeletePlayerAsync(PlayerDto.Delete player)
        {
            var exist = await players.SingleOrDefaultAsync(p => p.Id == player.Id);

            if (exist == null)
                return false;

            players.Remove(exist);

            return await context.SaveChangesAsync() > 0;
        }

        public async Task<IReadOnlyList<Player>> GetAllPlayers()
        {
            return await players
                .Include(p => p.Country)
                .AsNoTracking()
                .ToListAsync();
        }

        public async Task<IReadOnlyList<Player>> GetAllPlayersWithoutCountry()
        {
            return await players
                .AsNoTracking()
                .ToListAsync();
        }

        public async Task<Player> GetPlayerById(Guid id)
        {
            return await players
                .Include(p => p.Country)
                .SingleOrDefaultAsync(p => p.Id == id);
        }

        public async Task<Dictionary<string, double>> GetPercentageByCountry()
        {
            Dictionary<string, double> percentages = new();
            /*var query = players
                .Include(p => p.Country)
                .GroupBy(
                    p => p.Country.Code,
                    p => p,
                    (country, player) => new
                    {
                        Key = country,
                        Count = player.Count()
                    }
                );*/
            var query = players
                .Include(p => p.Country)
                .GroupBy(
                    p => new { p.Country.Code }
                )
                .Select(g => new
                {
                    Country = g.Key.Code,
                    Count = (double)g.Count() / players.Count() * 100
                }).OrderByDescending(x => x.Count);

            foreach (var result in query)
            {
                percentages.Add(result.Country, result.Count);
            }

            return percentages;
        }

        public async Task<bool> UpdatePlayerAsync(PlayerDto.Update player)
        {
            if (player == null)
                return false;

            var playerToUpdate = await players.Include(p => p.Country)
                .SingleOrDefaultAsync(p => p.Id == player.Id);

            if (playerToUpdate != null)
            {
                playerToUpdate.Name = player.Name;
                playerToUpdate.Active = player.Active;
            }

            return await context.SaveChangesAsync() > 0;
        }
    }
}

File: nodeBackEnd/services/playerService.ts
import { Prisma, PrismaClient } from "@prisma/client";
import { Player } from "../Models/player";
import { v4 as uuidv4 } from "uuid";
import { info } from "console";
import { PlayerDTOCreate, PlayerDTOUpdate } from "../Models/playerDTO";



export const getAllPlayers = async () => { 
    const prisma = new PrismaClient()
    const players =  await prisma.player.findMany({
        include: {
            Country: true
        },
    })

    return players
}

export const getAllPlayersWithoutCountry = async () => {
    const prisma = new PrismaClient()
    return await prisma.player.findMany()
} 

export const getPlayerById = async (playerId: string) => {    
    const prisma = new PrismaClient()
    const player = await prisma.player.findUnique({
        //include: { Country: true },
        where: { id: `${playerId}` },
    })
    
    return player
}

export const getPercentageByCountry = async() => {
    const prisma = new PrismaClient()
    const players = await prisma.player.findMany({
        include: { Country: true },
    })

    const percentages = new Map()
    players.map((player) => {
        if(percentages.has(player.Country?.code)) {
            let oldValue = percentages.get(player.Country?.code) as number
            percentages.set(player.Country?.code, oldValue + 1)
        }
        else {
            percentages.set(player.Country?.code, 1)
        }
    })

    for (let entry of Array.from(percentages.entries())) {
        percentages.set(entry[0], entry[1]/players.length * 100)
    }

    return percentages
}

export const updatePlayer = async(player: PlayerDTOUpdate) => {
    const prisma = new PrismaClient()
    let updatedPlayer = await prisma.player.update({
        where: { id: `${player.id}` },
        data: {
            name: `${player.name}`,
            active: player.active
        }
    })

    return updatedPlayer !== null
}

export const createPlayer = async(player: PlayerDTOCreate) => {
    const prisma = new PrismaClient()
    const existingPlayer = await prisma.player.findFirst({
        include: { Country: true },
        where: { 
            name: player.name,
            Country: {
                code: player.countryCode
            }
        }
    })

    if (existingPlayer === null) {
        let newPlayer: Player = {
            id: uuidv4(),
            playerId: player.playerId,
            name: player.name,
            gender: player.gender,
            active: player.active,
            birthyear: player.birthyear,
            playHand: player.playHand,
            playStyle: player.playStyle,
            grip: player.grip
        }

        const existingCountry = await prisma.country.findFirst({
            where: {
                code: player.countryCode
            }
        })

        let countryCode = player.countryCode
        if (existingCountry === null) {

            let newCountry = await prisma.country.create({
                data: {
                    id: uuidv4(),
                    code: countryCode
                },
                select: {
                    id: true,
                    code: true
                }
            })

            newPlayer.country = {
                id: newCountry.id,
                code: newCountry.code ?? undefined
            }
        } else {
            newPlayer.country = {
                id: existingCountry.id,
                code: existingCountry.code ?? undefined
            }
        }

        return await prisma.player.create({
            data: {
                id: newPlayer.id!,
                playerId: newPlayer.playerId,
                name: newPlayer.name,
                gender: newPlayer.gender,
                active: newPlayer.active,
                birthyear: newPlayer.birthyear,
                playHand: newPlayer.playHand,
                playStyle: newPlayer.playStyle,
                grip: newPlayer.grip,
                countryId: newPlayer.country.id
            }
        })
    }

    
}

export const deletePlayer = async(playerId: string) => {
    try {
        const prisma = new PrismaClient()
        let player = await prisma.player.findFirst({
            where: { id: playerId }
        })

        if(player !== null) {
            return await prisma.player.delete({
                where: { id: playerId }
            })
        }

        //throw 'The given user does not exist'
    } catch (e) {
        if (e instanceof Prisma.PrismaClientKnownRequestError) {
            if (e.code == 'P2025') {
                console.log("The record to delete does not exist")
            }
        }

        throw e
    }
    
}

const playerWithCountry = Prisma.validator<Prisma.PlayerArgs>()({
    include: { Country: true }
})
type PlayerWithCountry = Prisma.PlayerGetPayload<typeof playerWithCountry>
// npx prisma generate
